bindings/rust/src/lazy_reader.rs

```diff
@@ -0,0 +1,518 @@
+extern crate alloc;
+
+use alloc::boxed::Box;
+use alloc::format;
+use alloc::rc::Rc;
+use alloc::string::String;
+use alloc::vec;
+use alloc::vec::Vec;
+use core::cell::RefCell;
+use core::cmp::min;
+use core::convert::{From, TryFrom};
+
+#[derive(Debug)]
+pub enum Error {
+    Common(String),
+    TotalSize(String),
+    Header(String),
+    Offset(String),
+    UnknownItem(String),
+    OutOfBound(String),
+    FieldCount(String),
+    Data(String),
+    Overflow(String),
+    Read(String),
+    Verify(String),
+}
+pub trait Read {
+    /**
+     * try to read `buf.len()` bytes from data source with `offset`, then fill it in `buf`.
+     * the return size can be smaller than `buf.len()` which means the remaining data length is
+     * smaller than `buf.len()`
+     */
+    fn read(&self, buf: &mut [u8], offset: usize) -> Result<usize, Error>;
+}
+
+pub const MAX_CACHE_SIZE: usize = 2048;
+pub const MIN_CACHE_SIZE: usize = 64;
+pub const NUMBER_SIZE: usize = 4;
+
+pub struct DataSource {
+    reader: Box<dyn Read>,
+
+    total_size: usize,
+    cache_start_point: usize,
+    // cache size may be smaller than cache.len()
+    cache_size: usize,
+    cache: Vec<u8>,
+}
+
+#[derive(Clone)]
+pub struct Cursor {
+    pub offset: usize,
+    pub size: usize,
+    pub data_source: Rc<RefCell<DataSource>>,
+}
+
+pub struct Union {
+    pub item_id: usize,
+    pub cursor: Cursor,
+}
+
+impl Cursor {
+    /**
+    cache_size: normally it can be set to MAX_CACHE_SIZE(2K)
+    total_size: the size of cursor. If it's set a smaller value,
+    `out of bound` will occur when `reader` try to read the data beyond that.
+    reader: interface to read underlying data
+     */
+    pub fn new(total_size: usize, reader: Box<dyn Read>) -> Self {
```

Comment:
Let's think about this API: it suits the case when you are initializing a Cursor from a `Vec<u8>` structure, but what if you are initializing in a smart contract, from a syscall?

We would need to first build a reader implementing Read trait by reading syscalls, however, to initialize a Cursor, you would also need `total_size`, which is doable by... doing a syscall.

Thinking it another way, a DataSource can well be initialized by a single syscall, which loads enough data for the cache field, and also gets the total size of data in IO.

By requiring `total_size` here, we first need a syscall to get the total size, then DataSource requires another syscall to fill in the cache part. We are wasting one syscall here.

This is worth thinking, when refactoring the code related to DataSource.

Reply:
The `total_size` is required for safety/security reason: we can always check its bounds.

Reply:
I'm not sure on this: why `total_size` enhances security?